[RISCV][TTI] Enable masked interleave access

Mel-Chen opened the pull request:
Now that support for masked loads/stores of interleave groups has landed, we can enable the loop vectorizer to generate masked interleave access where applicable.

This improves vectorization in several ways:
* Internal predication support: This enables interleave group vectorization for loops with internal control flow predication, provided all members of the group share the same predicate. Gaps in interleave groups are still not efficiently handled by masking, so masking for gaps remains disabled for now.
* Tail folding: This allows tail folding of loops with interleave groups by using masking. Without this, vectorized loops with interleaves would fall back to using separate gather/scatter accesses, which can be significantly less efficient.

"[RISCV][TTI] Enable masked interleave access for scalable vector (#149981)" was reverted by 5294793bdcf6ca142f7a0df897638bd4e85ed1a7 due to triggering an assertion. The issue has been addressed in the patch "[LV] Fix gap mask requirement for interleaved access (#151105)". On the other hand, this patch also enable fixed-length masked interleave access (#150624) since support for fixed-length has also been landed 992118cb4deab139ae384bb85f03225a9a21b008.

Mel-Chen committed:
Reland "[RISCV][TTI] Enable masked interleave access for scalable vector (#149981)"

Now that support for masked loads/stores of interleave groups has landed, we can enable the loop vectorizer to generate masked interleave access where applicable.

This improves vectorization in several ways:
* Internal predication support: This enables interleave group vectorization for loops with internal control flow predication, provided all members of the group share the same predicate. Gaps in interleave groups are still not efficiently handled by masking, so masking for gaps remains disabled for now.
* Tail folding: This allows tail folding of loops with interleave groups by using masking. Without this, vectorized loops with interleaves would fall back to using separate gather/scatter accesses, which can be significantly less efficient.
* Scalable vector support: Currently, only scalable vector types are supported for masked interleave lowering. Fixed-length vector support will be enabled in the future.

As interleave access is not yet supported with tail folding by EVL, that functionality is temporarily disabled. We are going to create another patch to support it.

Co-authored-by: Philip Reames <preames@rivosinc.com>

lukel97 commented on llvm/lib/Target/RISCV/RISCVTargetTransformInfo.cpp:
```diff
@@ -979,10 +979,12 @@ InstructionCost RISCVTTIImpl::getInterleavedMemoryOpCost(
+  // only support masking per-iteration (i.e. condition), not per-segment (i.e.
+  // gap).
+  // TODO: Support masked interleaved access for fixed length vector.
+  if ((isa<ScalableVectorType>(VecTy) || !UseMaskForCond) && !UseMaskForGaps &&
```
Fixed vectors should be supported now, I don't think we need this check: https://github.com/llvm/llvm-project/pull/150624/files

Mel-Chen replied:
Yes, I saw it. @preames Do you mind if I enable both scalable and fixed-length masked interleaved access in this patch?
305e75fa01e42d8902c8d86b851895cd845fbe1c

lukel97 commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -1378,7 +1378,9 @@ class LoopVectorizationCostModel {
     // FIXME: Investigate opportunity for fixed vector factor.
+    // FIXME: Support interleave accesses.
     bool EVLIsLegal = UserIC <= 1 && IsScalableVF &&
+                      !InterleaveInfo.hasGroups() &&
```
I don't think we need to bail on EVL anymore either, https://github.com/llvm/llvm-project/pull/150202 has landed and should fix the concern about the header mask

Mel-Chen replied:
Hmm, I can't really think of any more cases that might cause issues. Hopefully it's safe enough. (pray)
767de324e473592107d29941f087c2d1c317ea39

Mel-Chen committed:
Cherry-pick #150624

Mel-Chen committed:
Remove limitation for EVL

lukel97 reviewed (approved):
LGTM, I tested it out locally and can confirm its generating masked vlseg/vsseg as expected from the loop vectorizer, both for internal predication and with EVL tail folding.

I think the last two restrictions in the PR description just need updated now that they're removed.

Mel-Chen committed:
Merge branch 'main' into reland-enable-masked-interleave

Mel-Chen commented:
> LGTM, I tested it out locally and can confirm its generating masked vlseg/vsseg as expected from the loop vectorizer, both for internal predication and with EVL tail folding.
> 
> I think the last two restrictions in the PR description just need updated now that they're removed.

Updated the title and description. Rebased and waiting for the test result.

lukel97 commented:
This gives some nice improvements across llvm-test-suite, uuencode is 33% faster now and there's some other big improvements to TSVC: https://lnt.lukelau.me/db_default/v4/nts/821
